extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java

```diff
@@ -185,47 +195,43 @@ void validateConfigInjectionPoints(ValidationPhaseBuildItem validationPhase,
-        // IMPL NOTE: we do depend on BeanContainerBuildItem to make sure that the BeanDeploymentValidator finished its processing
-
-        // the non-primitive types need to be registered for reflection since Class.forName is used at runtime to load the class
-        for (ConfigPropertyBuildItem item : configProperties) {
-            Type requiredType = item.getPropertyType();
-            String propertyType = requiredType.name().toString();
-            if (requiredType.kind() != Kind.PRIMITIVE) {
-                reflectiveClass.produce(ReflectiveClassBuildItem.builder(propertyType).build());
-            }
-        }
+    @Record(STATIC_INIT)
+    void validateStaticInitConfigProperty(
+            ConfigRecorder recorder,
+            List<ConfigPropertyBuildItem> configProperties,
+            BuildProducer<ReflectiveClassBuildItem> reflectiveClass) {
 
-        Set<ConfigValidationMetadata> propertiesToValidate = new HashSet<>();
-        for (ConfigPropertyBuildItem configProperty : configProperties) {
-            String rawTypeName = configProperty.getPropertyType().name().toString();
-            List<String> actualTypeArgumentNames = Collections.emptyList();
-            if (configProperty.getPropertyType().kind() == Kind.PARAMETERIZED_TYPE) {
-                List<Type> argumentTypes = configProperty.getPropertyType().asParameterizedType().arguments();
-                actualTypeArgumentNames = new ArrayList<>(argumentTypes.size());
-                for (Type argumentType : argumentTypes) {
-                    actualTypeArgumentNames.add(argumentType.name().toString());
-                    if (argumentType.kind() != Kind.PRIMITIVE) {
-                        reflectiveClass.produce(ReflectiveClassBuildItem.builder(argumentType.name().toString())
-                                .build());
-                    }
-                }
+        recorder.validateConfigProperties(
+                configProperties.stream()
+                        .filter(p -> p.getExecutionMode().equals(ExecutionMode.STATIC_INIT))
+                        .map(p -> configPropertyToConfigValidation(p, reflectiveClass))
+                        .collect(toSet()));
+    }
 
-            }
-            propertiesToValidate.add(new ConfigValidationMetadata(configProperty.getPropertyName(),
-                    rawTypeName, actualTypeArgumentNames, configProperty.getDefaultValue()));
-        }
+    @BuildStep
+    @Record(RUNTIME_INIT)
+    void validateRuntimeConfigProperty(
+            ConfigRecorder recorder,
+            List<ConfigPropertyBuildItem> configProperties,
+            BuildProducer<ReflectiveClassBuildItem> reflectiveClass) {
 
-        recorder.validateConfigProperties(propertiesToValidate);
+        recorder.validateConfigProperties(
+                configProperties.stream()
+                        .map(p -> configPropertyToConfigValidation(p, reflectiveClass))
```

Comment:
Shouldn't we filter out `static_init` `ConfigPropertyBuildItem`s here? I mean we [always register](https://github.com/quarkusio/quarkus/pull/39233/files#diff-26708186baae4b2358b51f3bc3c5731740eff57a0a63fb3a150d15183e817f55R205) a `runtime` variant, or?